src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/RazorCodeDocument.cs

```diff
@@ -17,41 +17,108 @@ public sealed partial class RazorCodeDocument
+        _tagHelpers = tagHelpers;
+        _referencedTagHelpers = referencedTagHelpers;
+        _preTagHelperSyntaxTree = preTagHelperSyntaxTree;
+        _syntaxTree = syntaxTree;
+        _importSyntaxTrees = importSyntaxTrees;
+        _tagHelperContext = tagHelperContext;
+        _documentNode = documentNode;
+        _cSharpDocument = cSharpDocument;
+
+        _htmlDocument = new Lazy<RazorHtmlDocument>(() => RazorHtmlWriter.GetHtmlDocument(this));
     }
 
-    public static RazorCodeDocument Create(
-        RazorSourceDocument source,
+    private RazorCodeDocument With(
+        RazorSourceDocument? source = null,
+        ImmutableArray<RazorSourceDocument>? imports = null,
         RazorParserOptions? parserOptions = null,
-        RazorCodeGenerationOptions? codeGenerationOptions = null)
+        RazorCodeGenerationOptions? codeGenerationOptions = null,
+        TagHelperCollection? tagHelpers = null,
+        TagHelperCollection? referencedTagHelpers = null,
+        RazorSyntaxTree? preTagHelperSyntaxTree = null,
+        RazorSyntaxTree? syntaxTree = null,
+        ImmutableArray<RazorSyntaxTree>? importSyntaxTrees = null,
+        TagHelperDocumentContext? tagHelperContext = null,
+        DocumentIntermediateNode? documentNode = null,
+        RazorCSharpDocument? cSharpDocument = null)
     {
-        ArgHelper.ThrowIfNull(source);
+        if (isChanged(source, Source)
+            || (imports is not null)
+            || isChanged(codeGenerationOptions, CodeGenerationOptions)
+            || isChanged(tagHelpers, _tagHelpers)
+            || isChanged(referencedTagHelpers, _referencedTagHelpers)
+            || isChanged(preTagHelperSyntaxTree, _preTagHelperSyntaxTree)
+            || isChanged(syntaxTree, _syntaxTree)
+            || (importSyntaxTrees is not null)
+            || isChanged(tagHelperContext, _tagHelperContext)
+            || isChanged(documentNode, _documentNode)
+            || isChanged(cSharpDocument, _cSharpDocument))
+        {
+            return new RazorCodeDocument(
+                source ?? Source,
+                imports ?? Imports,
+                parserOptions ?? ParserOptions,
+                codeGenerationOptions ?? CodeGenerationOptions,
+                tagHelpers ?? _tagHelpers,
+                referencedTagHelpers ?? _referencedTagHelpers,
+                preTagHelperSyntaxTree ?? _preTagHelperSyntaxTree,
+                syntaxTree ?? _syntaxTree,
+                importSyntaxTrees ?? _importSyntaxTrees,
+                tagHelperContext ?? _tagHelperContext,
+                documentNode ?? _documentNode,
+                cSharpDocument ?? _cSharpDocument);
+        }
+        else
+        {
+            return this;
+        }
 
-        return new RazorCodeDocument(source, imports: [], parserOptions, codeGenerationOptions);
+        static bool isChanged<T>(T? @new, T old) where T : class? => @new is not null && !ReferenceEquals(@new, old);
```

Comment:
>@new is not null [](http://example.com/codeflow?start=69&length=16)

why have the null check here? Wouldn't this make it so if I change something to be null it wouldn't get detected?